internal/attestation/azure/snp/validator.go

```diff
@@ -36,249 +41,368 @@ type Validator struct {
-				}
-				if certVersion != int(report.CommittedTCB.TEE) {
-					return fmt.Errorf("bootloader version %d from report does not match VCEK version %d", int(report.CommittedTCB.TEE), certVersion)
-				}
-			}
-		// check SNP Firmware version
-		case "1.3.6.1.4.1.3704.1.3.3":
-			{
-				_, err := asn1.Unmarshal(extension.Value, &certVersion)
-				if err != nil {
-					return fmt.Errorf("unmarshalling snp version: %w", err)
-				}
-				if certVersion != int(report.CommittedTCB.SNP) {
-					return fmt.Errorf("bootloader version %d from report does not match VCEK version %d", int(report.CommittedTCB.SNP), certVersion)
-				}
-			}
-		// check microcode version
-		case "1.3.6.1.4.1.3704.1.3.8":
-			{
-				_, err := asn1.Unmarshal(extension.Value, &certVersion)
-				if err != nil {
-					return fmt.Errorf("unmarshalling microcode version: %w", err)
-				}
-				if certVersion != int(report.CommittedTCB.Microcode) {
-					return fmt.Errorf("bootloader version %d from report does not match VCEK version %d", int(report.CommittedTCB.Microcode), certVersion)
-				}
-			}
+// azureInstanceInfo contains the necessary information to establish trust in
+// an Azure CVM.
+type azureInstanceInfo struct {
+	// VCEK is the PEM-encoded VCEK certificate for the attestation report.
+	VCEK []byte
+	// CertChain is the PEM-encoded certificate chain for the attestation report.
+	CertChain []byte
+	// AttestationReport is the attestation report from the vTPM (NVRAM) of the CVM.
+	AttestationReport []byte
+	// RuntimeData is the Azure runtime data from the vTPM (NVRAM) of the CVM.
+	RuntimeData []byte
+	// MAAToken is the token of the MAA for the attestation report, used as a fallback
+	// if the IDKeyDigests cannot be verified.
+	MAAToken string
+}
+
+// attestationWithCerts returns a formatted version of the attestation report and its certificates from the instanceInfo.
+// if the VCEK certificate or the certificate chain is not present, the given getter is used to retrieve them
+// from the AMD KDS.
+func (a *azureInstanceInfo) attestationWithCerts(logger attestation.Logger, getter trust.HTTPSGetter) (*spb.Attestation, error) {
+	report, err := abi.ReportToProto(a.AttestationReport)
+	if err != nil {
+		return nil, fmt.Errorf("converting report to proto: %w", err)
+	}
+
+	// Product info as reported through CPUID[EAX=1]
+	sevProduct := abi.DefaultSevProduct()
+	productName := kds.ProductString(sevProduct)
+
+	att := &spb.Attestation{
+		Report:           report,
+		CertificateChain: &spb.CertificateChain{},
+		Product:          sevProduct,
+	}
+
+	// If the VCEK certificate is present, parse it and format it.
+	vcek, vcekParsingErr := a.parseVCEK()
+	if err != nil {
+		logger.Warnf("Error parsing VCEK: %v", vcekParsingErr)
+	}
+	if vcek != nil {
+		att.CertificateChain.VcekCert = vcek.Raw
+	}
+	// Otherwise, retrieve it from AMD KDS.
+	if att.CertificateChain.VcekCert == nil {
+		logger.Infof("VCEK certificate not present, falling back to retrieving it from AMD KDS")
```

Comment:
can you explain why we can't use the "fill in missing certs" mechanism of the library?

Reply:
The [method they expose in their library](https://github.com/google/go-sev-guest/blob/bab68c55ee83798dc313c70cf9d98b61055e6cb9/verify/verify.go#L630-L642) unfortunately always fills in all the certificates, no matter which are already present. There is an [unexposed method](https://github.com/google/go-sev-guest/blob/bab68c55ee83798dc313c70cf9d98b61055e6cb9/verify/verify.go#L577-L628) that would do the job though. I will try to make this a public method upstream

Reply:
Opened an issue for this: 
https://github.com/google/go-sev-guest/issues/77

Reply:
It's also invoked here
https://github.com/google/go-sev-guest/blob/bab68c55ee83798dc313c70cf9d98b61055e6cb9/verify/verify.go#L548-L557
Wouldn't that suffice for us?

Reply:
Actually, we require the ASK to be present before calling the `SnpAttestation` method, due to the trustedRoots thing